### PR TITLE
Change when/how EagerPipe is used

### PR DESCRIPTION
**Pull request by henriknyman**
Use a less pessimistic view of when eagerness is needed in the Cypher execution plan
(WIP) PR created to run all tests

### CODE REVIEW

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/addEagernessIfNecessary.scala [hunk 1]
```diff
@@ -47,42 +36,84 @@ object addEagernessIfNecessary extends (Pipe => Pipe) {
...
-      case writes: WritesNodes => writes
-    }
+  private def wouldInterfere(from: Effects, to: Effects): Boolean = {
+//    val nodesInterfere = nodesReadInterference(from, to)
+//    val relsInterfere = from.contains(ReadsRelationships) && to.contains(WritesRelationships)
+
+    // NOTE: Leaf effects will not be considered unless effects have first been flattened with Effects.regardlessOfLeafEffects
+    val fromWithoutLeafInfo = from.regardlessOfLeafEffects
+    val toWithoutLeafInfo = to.regardlessOfLeafEffects
+
```
Comment (henriknyman) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/addEagernessIfNecessary.scala:64]: Re-order with most common interference first, and short-circuit evaluation on first interference

**Comment by pontusmelke** [comment]
Clean up commit messages

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/Effects.scala [hunk 2]
```diff
@@ -149,59 +163,72 @@ object ReadsNodesWithLabels {
...
-case object WritesRelationshipBoundNodes extends WritesNodes
+trait CreatesNodes extends WritesNodes
 
-object WritesNodesWithLabels {
-  def apply(labels: String*): WritesNodesWithLabels = WritesNodesWithLabels(labels.toSet)
+case object CreatesAnyNode extends CreatesNodes
+
+case object CreatesRelationshipBoundNodes extends CreatesNodes
+
+//case object DeletesRelationshipBoundNodes
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/Effects.scala]: remove?

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/Effects.scala [hunk 3]
```diff
@@ -108,39 +125,36 @@ object Effects {
...
-  override def reads = true
+protected trait ReadEffect extends Effect
+
+protected trait WriteEffect extends Effect
 
-  override def writes = false
+case class LeafEffect(effect: Effect) extends Effect {
 
-  def toWriteEffect: WriteEffect
+  override def toString = s"${super.toString} (${effect.toString})"
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/Effects.scala]: Why are we overriding `toString`, it doesn't give us anything that is not already in the default case class `toString`? I vote for removing the toString of `Effect` and just use default.

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/Effects.scala [hunk 4]
```diff
@@ -149,59 +163,72 @@ object ReadsNodesWithLabels {
...
+  def apply(labels: String*): CreatesNodesWithLabels = CreatesNodesWithLabels(labels.toSet)
+}
 
 trait ReadsNodeProperty extends ReadEffect
 
 case class ReadsGivenNodeProperty(propertyName: String) extends ReadsNodeProperty {
-  override def toString = s"${super.toString} '$propertyName'"
 
-  override def toWriteEffect = WritesGivenNodeProperty(propertyName)
+  override def toString = s"${super.toString} '$propertyName'"
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/Effects.scala]: as above

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/Effects.scala [hunk 5]
```diff
@@ -149,59 +163,72 @@ object ReadsNodesWithLabels {
...
 
-  override def toString = this.getClass.getSimpleName
-}
-
-trait WritesNodeProperty extends WriteEffect
+trait SetNodeProperty extends WriteEffect
 
-case class WritesGivenNodeProperty(propertyName: String) extends WritesNodeProperty {
+case class SetGivenNodeProperty(propertyName: String) extends SetNodeProperty {
   override def toString = s"${super.toString} '$propertyName'"
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/Effects.scala]: as above

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/Effects.scala [hunk 6]
```diff
@@ -149,59 +163,72 @@ object ReadsNodesWithLabels {
...
-object WritesAnyNodeProperty extends WritesNodeProperty {
-  override def toString = this.getClass.getSimpleName
+object SetAnyNodeProperty extends SetNodeProperty
+
+trait ReadsRelationships extends ReadEffect
+
+case object ReadsAllRelationships extends ReadsRelationships
+
+case class ReadsRelationshipsWithTypes(types: Set[String]) extends ReadsRelationships {
+  override def toString = s"${super.toString} '${types.mkString(", ")}'"
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/Effects.scala]: as above

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/Effects.scala [hunk 7]
```diff
@@ -149,59 +163,72 @@ object ReadsNodesWithLabels {
...
+    else ReadsRelationshipsWithTypes(types.toSet)
 }
 
-case object WritesRelationships extends WriteEffect
+case object DeletesRelationship extends WriteEffect
+
+trait CreatesRelationships extends WriteEffect
+
+case class CreatesRelationship(typ: String) extends CreatesRelationships {
+  override def toString = s"${super.toString} '$typ'"
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/Effects.scala]: as above

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/Effects.scala [hunk 8]
```diff
@@ -149,59 +163,72 @@ object ReadsNodesWithLabels {
...
+}
 
 trait ReadsRelationshipProperty extends ReadEffect
 
 case class ReadsGivenRelationshipProperty(propertyName: String) extends ReadsRelationshipProperty {
   override def toString = s"${super.toString} '$propertyName'"
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/Effects.scala]: as above

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/Effects.scala [hunk 9]
```diff
@@ -209,6 +236,4 @@ case class WritesGivenRelationshipProperty(propertyName: String) extends WritesR
...
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/Effects.scala]: as above

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/addEagernessIfNecessary.scala [hunk 10]
```diff
@@ -22,17 +22,7 @@ package org.neo4j.cypher.internal.compiler.v2_3.executionplan
...
-    val relsInterfere = from.contains(ReadsRelationships) && to.contains(WritesRelationships)
-
-    val readWriteInterfereNodes = nodesWriteInterference(from, to)
-    val readWriteInterfereRelationships = from.contains(WritesRelationships) && to.contains(WritesRelationships) &&
-      to.contains(ReadsRelationships)
-
-    nodesInterfere || relsInterfere || readWriteInterfereNodes || readWriteInterfereRelationships ||
-      nodePropertiesInterfere(from, to) || relationshipPropertiesInterfere(from, to)
-  }
+  private val DEBUG = true
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/addEagernessIfNecessary.scala]: Should this be true?

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/Effects.scala (hunk 3 above)
Comment (henriknyman) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/Effects.scala]: Good point

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/OperationsFacade.java [hunk 11]
```diff
@@ -325,6 +325,13 @@ public RelationshipIterator nodeGetRelationships( long nodeId, Direction directi
             throws EntityNotFoundException
     {
         statement.assertOpen();
+
+        // TODO: Is it ok to handle this case here?
```
Comment (henriknyman) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/OperationsFacade.java]: Should we get an opinion from somebody in the kernel team on this?

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/pipes/EagerPipe.scala [hunk 12]
```diff
@@ -37,8 +37,11 @@ case class EagerPipe(src: Pipe)(implicit pipeMonitor: PipeMonitor) extends PipeW
 
   def planDescription = src.planDescription.andThen(this.id, "Eager", identifiers)
 
-  protected def internalCreateResults(input: Iterator[ExecutionContext], state: QueryState): Iterator[ExecutionContext] =
-    input.toList.toIterator
+  protected def internalCreateResults(input: Iterator[ExecutionContext], state: QueryState): Iterator[ExecutionContext] = {
+    val list: List[ExecutionContext] = input.toList
+    println(s"the eager iterator pulled in ${list.size} elements")
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/pipes/EagerPipe.scala]: remove

File: community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/AddEagernessIfNecessaryTest.scala [hunk 13]
```diff
@@ -62,100 +62,118 @@ class AddEagernessIfNecessaryTest extends CypherFunSuite {
...
-    .to(Effects(WritesAnyNode))
+    testThatGoingFrom(Effects(ReadsAllRelationships))
+    .to(Effects(CreatesAnyNode))
     .doesNotIntroduceEagerness()
   }
 
-  test("READS PROP -> WRITES PROP needs eagerness"){
-    testThatGoingFrom(Effects(ReadsGivenNodeProperty("foo")))
-      .to(Effects(WritesGivenNodeProperty("foo")))
+//  test("READS PROP -> WRITES PROP needs eagerness"){
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/AddEagernessIfNecessaryTest.scala]: Fix commented out tests

File: community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/EffectsTest.scala [hunk 14]
```diff
@@ -17,87 +17,106 @@
...
-  }
-
-  test("logical OR considers equal label names") {
-    val first = Effects(WritesAnyNode, ReadsNodesWithLabels("bar"))
-    val second = Effects(ReadsNodesWithLabels("bar"))
-
-    (first | second) should be(Effects(WritesAnyNode, ReadsNodesWithLabels("bar")))
-  }
-}
+///*
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/EffectsTest.scala]: Fix commented out stuff

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala [hunk 15]
```diff
@@ -47,615 +90,721 @@ class EagerizationAcceptanceTest extends ExecutionEngineFunSuite with TableDrive
...
-  }
+  test("should not need eagerness for match create with labels") {
+    createLabeledNode("L")
+    val query = "MATCH (:L) CREATE (:L)"
 
-  test("MATCH (a:Person),(m:Movie) CREATE (a)-[:T]->(m) WITH a OPTIONAL MATCH (a) RETURN *") {
-    val query = "MATCH (a:Person),(m:Movie) CREATE (a)-[:T]->(m) WITH a OPTIONAL MATCH (a) RETURN *"
+    val result = execute(query)
 
+    println(result.dumpToString())
```
Comment (pontusmelke) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala]: remove

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala [hunk 16]
```diff
@@ -47,615 +90,721 @@ class EagerizationAcceptanceTest extends ExecutionEngineFunSuite with TableDrive
...
-  test("should not add eagerness when reading and merging nodes and relationships when matching different label") {
-    val query = "MATCH (a:A) MERGE (a)-[:BAR]->(b:B) WITH a MATCH (a) WHERE (a)-[:FOO]->() RETURN a"
+  test("should not need eagerness for match create with labels and property with index") {
+    createLabeledNode(Map("id" -> 0), "L")
+    graph.createIndex("L", "id")
+    val query = "MATCH (:L {id: 0}) CREATE (:L {id:0})"
+
+    val result = execute(query)
 
+    println(result.dumpToString())
```
Comment (pontusmelke) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala]: remove

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala [hunk 17]
```diff
@@ -47,615 +90,721 @@ class EagerizationAcceptanceTest extends ExecutionEngineFunSuite with TableDrive
...
-  test("should add eagerness when reading and merging nodes and relationships on matching same label") {
-    val query = "MATCH (a:A) MERGE (a)-[:BAR]->(b:A) WITH a MATCH (a) WHERE (a)-[:FOO]->() RETURN a"
+  test("should need eagerness for double match and then create") {
+    createNode()
+    createNode()
+    val query = "MATCH (), () CREATE ()"
 
+    val result = execute(query)
+
+    println(result.dumpToString())
```
Comment (pontusmelke) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala]: remove

File: community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/EffectsTest.scala [hunk 18]
```diff
@@ -1,103 +0,0 @@
-/*
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/EffectsTest.scala]: Isn't Effects a big enough thing to merit at least some unit tests?

**Comment by systay** [comment]
We usually rebase PRs and don't use merge commits like this. Was this done intentionally to start a discussion about this practise, or can we just rebase it?

**Comment by Mats-SX** [comment]
Rebased and removed merge commit.

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/addEagernessIfNecessary.scala (hunk 1 above)
Comment (srbaker) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/addEagernessIfNecessary.scala]: Please don't commit the commented out code. We've got version control for this! :)

**Pull request merged**